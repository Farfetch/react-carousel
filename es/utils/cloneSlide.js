function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { cloneElement } from 'react';
import Styles from "../components/Slider.css";
import cx from 'classnames';
import memoize from 'fast-memoize';

var getElementStyle = function getElementStyle(style, styleChild) {
  return _objectSpread(_objectSpread({}, style), styleChild);
};

var memoizedGetElementStyle = memoize(getElementStyle);

var cloneSlide = function cloneSlide(child, index, isCloned, className, style) {
  var newStyle = memoizedGetElementStyle(style, child.props.style);
  return /*#__PURE__*/cloneElement(child, {
    className: cx(className, Styles.slide, child.props.className),
    style: newStyle,
    key: isCloned ? "c.".concat(index) : child.key,
    ref: function ref(node) {
      // Keep ref when use cloneElement - Dan Abramov's tweet:
      // https://twitter.com/dan_abramov/status/752944645018189825?lang=en
      if (child.ref && !isCloned) {
        child.ref(node);
      }
    }
  });
};

export default cloneSlide;