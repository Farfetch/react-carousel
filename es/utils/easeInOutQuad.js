export default (function (t) {
  return t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t;
});