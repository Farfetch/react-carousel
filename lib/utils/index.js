"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "checkSnapSupport", {
  enumerable: true,
  get: function get() {
    return _checkSnapSupport["default"];
  }
});
Object.defineProperty(exports, "cloneSlide", {
  enumerable: true,
  get: function get() {
    return _cloneSlide["default"];
  }
});
Object.defineProperty(exports, "renderSlides", {
  enumerable: true,
  get: function get() {
    return _renderSlides["default"];
  }
});
Object.defineProperty(exports, "animate", {
  enumerable: true,
  get: function get() {
    return _animate["default"];
  }
});
Object.defineProperty(exports, "animateScroll", {
  enumerable: true,
  get: function get() {
    return _animate.animateScroll;
  }
});
Object.defineProperty(exports, "easeInOutQuad", {
  enumerable: true,
  get: function get() {
    return _easeInOutQuad["default"];
  }
});
Object.defineProperty(exports, "infiniteScroll", {
  enumerable: true,
  get: function get() {
    return _unevenInfiniteScroll["default"];
  }
});

var _checkSnapSupport = _interopRequireDefault(require("./checkSnapSupport"));

var _cloneSlide = _interopRequireDefault(require("./cloneSlide"));

var _renderSlides = _interopRequireDefault(require("./renderSlides"));

var _animate = _interopRequireWildcard(require("./animate"));

var _easeInOutQuad = _interopRequireDefault(require("./easeInOutQuad"));

var _unevenInfiniteScroll = _interopRequireDefault(require("./unevenInfiniteScroll"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }