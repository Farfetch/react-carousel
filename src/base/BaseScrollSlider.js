import { SCROLL, animateScroll, easeInOutQuad, renderSlides } from '#utils';
import PropTypes from 'prop-types';
import React, { Children, Component, createRef } from 'react';
import debounce from 'lodash/debounce';

class BaseScrollSlider extends Component {
    constructor(props) {
        super(props);
        this._verifyActiveItemDebounce = debounce(this._verifyActiveItem, 25);
        this._handleScroll = this._handleScroll.bind(this);
        this.itemsLength = Children.count(props.children);
        this.activeItem = 0;
        this.shouldMoveToItem = false;
        this.isPreventingScroll = false;
        this.containerRef = createRef();
    }

    componentDidMount() {
        this.props.setItemsLength(this.itemsLength);
    }

    componentDidUpdate() {
        if (this.shouldMoveToItem) {
            this._moveToItem();
        }
    }

    render() {
        const {
            activeItem,
            children,
            goTo,
            isRTL,
            itemsToShow,
            setItemsLength,
            animationDuration,
            animationTimingFunction,
            className = 'scrollSliderContainer',
            type,
            ...othersProps
        } = this.props;

        if (activeItem !== this.activeItem) {
            this.activeItem = activeItem;
            this.shouldMoveToItem = true;
        } else {
            this.shouldMoveToItem = false;
        }

        const style = {
            width: `${(100 / itemsToShow).toFixed(6)}%`,
        };

        return (
            <div
                className={className}
                ref={this.containerRef}
                onScroll={this._handleScroll}
                {...othersProps}
            >
                {renderSlides(children, itemsToShow, style)}
            </div>
        );
    }

    _moveToItem() {
        const { animationDuration, animationTimingFunction, type } = this.props;
        const el = this.containerRef.current;

        if (!el) {
            return null;
        }

        const { offsetWidth } = el;
        const { activeItem, itemsToShow, isRTL } = this.props;

        const itemToMoveTo = isRTL
            ? this.itemsLength - 2 - activeItem
            : activeItem;
        const finalPosition = (offsetWidth / itemsToShow) * itemToMoveTo;

        animateScroll(
            el,
            finalPosition,
            animationDuration,
            animationTimingFunction,
            type,
            () => {
                this.shouldMoveToItem = false;
            }
        );
    }

    _verifyActiveItem() {
        const { goTo, itemsToShow, isRTL } = this.props;
        const { offsetWidth, scrollLeft } = this.containerRef.current;

        const activeFraction =
            Math.round((scrollLeft / offsetWidth) * 100) / 100;
        const newActiveItem = isRTL
            ? this.itemsLength -
              1 -
              Math.round(activeFraction * itemsToShow) -
              (itemsToShow - 1)
            : Math.round(activeFraction * itemsToShow);

        if (this.activeItem === newActiveItem) {
            return null;
        }

        this.activeItem = newActiveItem;
        goTo(newActiveItem);
    }

    _handleScroll(e) {
        if (this.isPreventingScroll) {
            e.stopPropagation();
            e.preventDefault();

            return null;
        }

        if (this.shouldMoveToItem) {
            return null;
        }

        this._verifyActiveItemDebounce();
    }
}

BaseScrollSlider.defaultProps = {
    animationDuration: 0.3,
    animationTimingFunction: easeInOutQuad,
    type: SCROLL,
};

BaseScrollSlider.propTypes = {
    activeItem: PropTypes.number.isRequired,
    children: PropTypes.node.isRequired,
    isRTL: PropTypes.bool.isRequired,
    itemsToShow: PropTypes.number.isRequired,
    setItemsLength: PropTypes.func.isRequired,
    goTo: PropTypes.func.isRequired,
    animationDuration: PropTypes.number,
    animationTimingFunction: PropTypes.func,
    type: PropTypes.string,
    className: PropTypes.string,
};

export default BaseScrollSlider;
