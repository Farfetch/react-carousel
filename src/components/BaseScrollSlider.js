/* import { CarouselContext } from '#context';
import { animateScroll, quad, renderSlides } from '#utils';
import PropTypes from 'prop-types';
import React, { forwardRef, useContext } from 'react';

const BaseScrollSlider = forwardRef(({ children, as: As, ...rest }, ref) => {
    const {
        activeItem,
        isRTL,
        itemsToShow,
        setItemsLength,
        goTo,
    } = React.useContext(CarouselContext);

    return (
        <As ref={ref} {...rest}>
            {children}
        </As>
    );
});

BaseScrollSlider.displayName = 'BaseScrollSlider';

BaseScrollSlider.defaultProps = {
    as: 'div',
};

BaseScrollSlider.propTypes = {
    as: PropTypes.string,
    children: PropTypes.element,
};

export default BaseScrollSlider;
 */
