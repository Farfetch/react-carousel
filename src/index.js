export { checkSnapSupport } from '#utils';
export { CarouselContext, CarouselProvider } from '#context';
export { Arrow, BaseArrow, Carousel, SnapSlider } from '#components';
export {
    Bullets,
    SwipeSlider,
    ScrollSlider,
    UnevenItemsScrollSlider,
} from '#containers';
