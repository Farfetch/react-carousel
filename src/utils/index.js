export { default as checkSnapSupport } from './checkSnapSupport';
export { default as cloneSlide } from './cloneSlide';
export { default as renderSlides } from './renderSlides';
export { default as animate } from './animate';
export { animateScroll } from './animate';
export { default as easeInOutQuad } from './easeInOutQuad';
export { default as infiniteScroll } from './unevenInfiniteScroll';
