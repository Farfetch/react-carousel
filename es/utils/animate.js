import easeInOutQuad from "./easeInOutQuad";

var animate = function animate(callbackObj, durationInSec) {
  var durationInMs = durationInSec * 1000 || 1000;
  var startTime = 0;
  var percentage = 0;
  var animationTime = 0;

  var animation = function animation(timestamp) {
    if (startTime === 0) {
      startTime = timestamp;
    } else {
      animationTime = timestamp - startTime;
    }

    if (typeof callbackObj.start === 'function' && startTime === timestamp) {
      callbackObj.start();
      window.requestAnimationFrame(animation);
    } else if (animationTime < durationInMs) {
      if (typeof callbackObj.progress === 'function') {
        percentage = animationTime / durationInMs;
        callbackObj.progress(percentage);
      }

      window.requestAnimationFrame(animation);
    } else if (typeof callbackObj.done === 'function') {
      callbackObj.done();
    }
  };

  return window.requestAnimationFrame(animation);
};

export var animateScroll = function animateScroll(el, finalPosition, transitionDuration, onDone) {
  var startPosition = el.scrollLeft;
  var diffPosition = finalPosition - startPosition;
  var callback = {
    progress: function progress(percentage) {
      var animationPosition = startPosition + easeInOutQuad(percentage) * diffPosition;

      if (el.scroll) {
        el.scroll(animationPosition, 0);
      } else {
        // IE/Edge do not support scroll() or scrollTo()
        el.scrollLeft = animationPosition;
      }
    },
    done: function done() {
      if (el.scroll) {
        el.scroll(startPosition + diffPosition, 0);
      } else {
        // IE/Edge do not support scroll() or scrollTo()
        el.scrollLeft = startPosition + diffPosition;
      }

      if (typeof onDone === 'function') {
        onDone();
      }
    }
  };
  animate(callback, transitionDuration);
};
export default animate;