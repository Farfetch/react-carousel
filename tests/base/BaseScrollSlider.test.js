import { animateScroll } from '../../src/utils';
import { createSerializer } from 'enzyme-to-json';
import { shallow } from 'enzyme';
import BaseScrollSlider from '../../src/base/BaseScrollSlider';
import React from 'react';

jest.mock('../../src/utils/animate', () => ({
    __esModule: true,
    default: jest.fn((callback) => {
        callback.progress();
        callback.done();
    }),
    animateScroll: jest.fn(),
}));

const children = [
    <div key="0">First slide</div>,
    <div key="1">Second slide</div>,
    <div key="2">Third slide</div>,
    <div key="3">Fourth slide</div>,
];

const mockRequiredProps = {
    activeItem: 0,
    children,
    isRTL: false,
    itemsToShow: 2,
    setItemsLength: jest.fn(),
    goTo: jest.fn(),
};

describe('<BaseScrollSlider />', () => {
    beforeAll(() => {
        jest.useFakeTimers();
        expect.addSnapshotSerializer(createSerializer());
    });

    afterEach(() => {
        jest.clearAllMocks();
        jest.clearAllTimers();
    });

    describe('<BaseScrollSlider /> - props', () => {
        it('should render correctly with the required props', () => {
            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            expect(wrapper.children()).toHaveLength(4);
            expect(wrapper).toMatchSnapshot();
        });

        it('should render correctly with aditional props', () => {
            const mockProps = {
                ...mockRequiredProps,
                limitScroll: true,
                className: 'dummyClassName',
            };
            const wrapper = shallow(<BaseScrollSlider {...mockProps} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('<ScrollSlider> - lifecycling', () => {
        it('should call setItemsLength after first mount', () => {
            shallow(<BaseScrollSlider {...mockRequiredProps} />, {
                disableLifecycleMethods: false,
            });

            expect(mockRequiredProps.setItemsLength).toHaveBeenCalledTimes(1);
            expect(mockRequiredProps.setItemsLength).toHaveBeenCalledWith(4);
        });

        it('should call moveToItem when active item changes', () => {
            const spyMoveToItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_moveToItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />,
                {
                    disableLifecycleMethods: false,
                }
            );

            wrapper.setProps({ activeItem: 1 });

            expect(spyMoveToItem).toHaveBeenCalledTimes(1);
        });

        it('should not call moveToItem when active item does not change', () => {
            const spyMoveToItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_moveToItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />,
                {
                    disableLifecycleMethods: false,
                }
            );

            wrapper.setProps({ activeItem: 0 });

            expect(spyMoveToItem).toHaveBeenCalledTimes(0);
        });
    });

    describe('_moveToItem', () => {
        it('should return null when current containerRef is not defined', () => {
            const spyMoveToItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_moveToItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            const instance = wrapper.instance();

            instance._moveToItem();

            expect(spyMoveToItem).toHaveReturnedWith(null);
            expect(animateScroll).toHaveBeenCalledTimes(0);
        });

        it('should request animation frame when containerRef is defined', () => {
            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            const instance = wrapper.instance();

            instance.containerRef.current = {
                offsetWidth: 100,
                scrollLeft: 1,
            };

            instance._moveToItem();

            expect(animateScroll).toHaveBeenCalledTimes(1);
        });

        it('should request animation frame when containerRef is defined and is rtl', () => {
            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} isRTL />
            );

            const instance = wrapper.instance();

            instance.containerRef.current = {
                offsetWidth: 100,
                scrollLeft: 1,
            };

            instance._moveToItem();

            expect(animateScroll).toHaveBeenCalledTimes(1);
        });
    });

    describe('_verifyActiveItem', () => {
        it('should return null when active item does not change', () => {
            const spyVerifyActiveItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_verifyActiveItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            const instance = wrapper.instance();

            instance.containerRef.current = {
                offsetWidth: 100,
                scrollLeft: 1,
            };

            instance._verifyActiveItem();

            expect(spyVerifyActiveItem).toHaveReturnedWith(null);
            expect(mockRequiredProps.goTo).toHaveBeenCalledTimes(0);
        });

        it('should call goTo and set new active item correctly when active item changes', () => {
            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            const instance = wrapper.instance();

            instance.containerRef.current = {
                offsetWidth: 100,
                scrollLeft: 51,
            };

            instance._verifyActiveItem();

            expect(mockRequiredProps.goTo).toHaveBeenCalledTimes(1);
            expect(mockRequiredProps.goTo).toHaveBeenCalledWith(1);
        });

        it('should call goTo and set new active item correctly when active item changes and is rtl', () => {
            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} isRTL />
            );

            const instance = wrapper.instance();

            instance.containerRef.current = {
                offsetWidth: 100,
                scrollLeft: 51,
            };

            instance._verifyActiveItem();

            expect(mockRequiredProps.goTo).toHaveBeenCalledTimes(1);
            expect(mockRequiredProps.goTo).toHaveBeenCalledWith(1);
        });
    });

    describe('_handleScroll', () => {
        it('should return null when shouldMoveToItem is already true', () => {
            const spyHandleScroll = jest.spyOn(
                BaseScrollSlider.prototype,
                '_handleScroll'
            );
            const spyVerifyActiveItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_verifyActiveItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            wrapper.setProps({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._handleScroll();

            expect(instance.shouldMoveToItem).toBe(true);
            expect(spyHandleScroll).toHaveReturnedWith(null);
            expect(spyVerifyActiveItem).toHaveBeenCalledTimes(0);
        });

        it('should return null when scroll is locked', () => {
            const spyHandleScroll = jest.spyOn(
                BaseScrollSlider.prototype,
                '_handleScroll'
            );
            const spyVerifyActiveItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_verifyActiveItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            const instance = wrapper.instance();

            instance.isPreventingScroll = true;

            const event = {
                stopPropagation: jest.fn(),
                preventDefault: jest.fn(),
            };

            instance._handleScroll(event);

            expect(event.preventDefault).toHaveBeenCalledTimes(1);
            expect(event.stopPropagation).toHaveBeenCalledTimes(1);
            expect(spyHandleScroll).toHaveReturnedWith(null);
            expect(spyVerifyActiveItem).toHaveBeenCalledTimes(0);
        });

        it('should verify active item when limit scroll is not active', () => {
            const spyVerifyActiveItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_verifyActiveItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} />
            );

            const instance = wrapper.instance();

            instance.containerRef.current = {
                offsetWidth: 100,
                scrollLeft: 100,
            };

            instance._handleScroll();

            jest.runAllTimers();

            expect(spyVerifyActiveItem).toHaveBeenCalledTimes(1);
        });

        it('should verify active item when limit scroll is defined and scroll does not exceed limit', () => {
            const spyVerifyActiveItem = jest.spyOn(
                BaseScrollSlider.prototype,
                '_verifyActiveItem'
            );

            const wrapper = shallow(
                <BaseScrollSlider {...mockRequiredProps} limitScroll />
            );

            const instance = wrapper.instance();

            instance.containerRef.current = {
                offsetWidth: 100,
                scrollLeft: 1,
            };

            instance._handleScroll();

            jest.runAllTimers();

            expect(spyVerifyActiveItem).toHaveBeenCalledTimes(1);
        });
    });
});
