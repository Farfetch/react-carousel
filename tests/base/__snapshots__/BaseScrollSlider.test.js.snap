// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseScrollSlider /> <BaseScrollSlider /> - props should render correctly with aditional props 1`] = `
<div
  className="dummyClassName"
  limitScroll={true}
  onScroll={[Function]}
>
  <div
    key=".$0"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    First slide
  </div>
  <div
    key=".$1"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    Second slide
  </div>
  <div
    key=".$2"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    Third slide
  </div>
  <div
    key=".$3"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    Fourth slide
  </div>
</div>
`;

exports[`<BaseScrollSlider /> <BaseScrollSlider /> - props should render correctly with the required props 1`] = `
<div
  onScroll={[Function]}
>
  <div
    key=".$0"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    First slide
  </div>
  <div
    key=".$1"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    Second slide
  </div>
  <div
    key=".$2"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    Third slide
  </div>
  <div
    key=".$3"
    style={
      Object {
        "width": "50.000000%",
      }
    }
  >
    Fourth slide
  </div>
</div>
`;
